File: MmoGameUnity/Assets/Game/Net/Net/CsProtocol/Protocol/Bag/Item/ItemBaseData.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using ZJYFrameWork.Net.Core;

namespace ZJYFrameWork.Net.CsProtocol.Buffer.Protocol.Bag.Item
{
    public class ItemBaseData : Model, IPacket
    {
        public int id;
        public string name;
        public string icon;
        public int minNum;
        public int maxNum;
        public int type;
        public string des;

        public static ItemBaseData ValueOf()
        {
            return new ItemBaseData();
        }

        public static ItemBaseData ValueOf(string des, string icon, int id, int maxNum, int minNum, string name,
            int type)
        {
            var packet = new ItemBaseData
            {
                des = des,
                icon = icon,
                id = id,
                maxNum = maxNum,
                minNum = minNum,
                name = name,
                type = type
            };
            return packet;
        }


        public short ProtocolId()
        {
            return 201;
        }
    }

    public class ItemBaseDataRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 201;
        }

        public void Write(ByteBuffer buffer, IPacket packet)
        {
            ItemBaseData message = (ItemBaseData)packet;
            var _message = new ServerMessageWrite(message.ProtocolId(), message);
            var json = JsonConvert.SerializeObject(_message);
            buffer.WriteString(json);
        }

        public IPacket Read(ByteBuffer buffer, Dictionary<object, object> dict)
        {
            dict.TryGetValue("packet", out var packetJson);
            if (packetJson != null)
            {
                var packet = JsonConvert.DeserializeObject<ItemBaseData>(packetJson.ToString());

                return packet;
            }

            return null;
        }
    }
}

File: DemoMMOUnity/Lua/application/app/net/webscoket/luaProtocol/bag/itemBaseData.lua
---@class ItemBaseData
local ItemBaseData = class("ItemBaseData")
local this = ItemBaseData
function ItemBaseData:ctor()
    ---@type number
    self.id = 0
    ---@type string
    self.name = string.empty
    ---@type string
    self.icon = string.empty
    ---@type number
    self.minNum = 0
    ---@type number
    self.maxNum = 0
    ---@type number
    self.type = 0
    ---@type string
    self.des = string.empty
end

---@param id integer
---@param name string
---@param icon string
---@param minNum integer
---@param maxNum integer
---@param type integer
---@param des string
function ItemBaseData:new(id,
                          name,
                          icon,
                          minNum,
                          maxNum,
                          type,
                          des)
    self.id = id
    self.name = name
    self.icon = icon
    self.minNum = minNum
    self.maxNum = maxNum
    self.type = type
    self.des = des
    return this
end

function ItemBaseData:protocolId()
    return 201
end

function ItemBaseData:write(buffer, packet)
    if packet == nil then
        return
    end
    local data = packet or ItemBaseData
    local message = {
        protocolId = data.protocolId(),
        packet = { id = self.id, name = self.name, icon = self.icon, minNum = self.minNum, maxNum = self.maxNum, type = self.type, des = self.des }
    }
    local jsonStr = JSON.encode(message)
    buffer:writeString(jsonStr)
end

function ItemBaseData:read(data)
    return ItemBaseData:new(
        data.id,
        data.name,
        data.icon,
        data.minNum,
        data.maxNum,
        data.type,
        data.des)
end

return ItemBaseData
